s3/src/main/java/ch/cyberduck/core/sts/STSAssumeRoleAuthorizationService.java

```
@@ -0,0 +1,163 @@
+        final String token;
+        if(StringUtils.isNotBlank(oauth.getIdToken())) {
+            if(log.isDebugEnabled()) {
+                log.debug(String.format("Assume role with OIDC Id token for %s", bookmark));
+            }
+            token = oauth.getIdToken();
+        }
+        else {
+            if(log.isDebugEnabled()) {
+                log.debug(String.format("Assume role with OAuth access token for %s", bookmark));
+            }
+            token = oauth.getAccessToken();
+        }
+        request.setWebIdentityToken(token);
+        final HostPreferences preferences = new HostPreferences(bookmark);
+        if(preferences.getInteger("s3.assumerole.durationseconds") != 0) {
+            request.setDurationSeconds(preferences.getInteger("s3.assumerole.durationseconds"));
+        }
+        if(StringUtils.isNotBlank(preferences.getProperty("s3.assumerole.policy"))) {
+            request.setPolicy(preferences.getProperty("s3.assumerole.policy"));
+        }
+        if(StringUtils.isNotBlank(preferences.getProperty("s3.assumerole.rolearn"))) {
+            request.setRoleArn(preferences.getProperty("s3.assumerole.rolearn"));
+        }
+        if(StringUtils.isNotBlank(preferences.getProperty("s3.assumerole.rolesessionname"))) {
+            request.setRoleSessionName(preferences.getProperty("s3.assumerole.rolesessionname"));
+        }
+        else {
+            try {
+                final String sub = JWT.decode(token).getSubject();
+                if(StringUtils.isNotBlank(sub)) {
+                    request.setRoleSessionName(sub);
+                }
+                else {
+                    log.warn(String.format("Missing subject in decoding JWT %s", token));
+                    request.setRoleSessionName(new AsciiRandomStringService().random());
+                }
+            }
+            catch(JWTDecodeException e) {
+                log.warn(String.format("Failure decoding JWT %s", token));
```

@chenkins We should probably fail fast here with an exception if the token cannot be decoded as a JWT?

---

In 204949f8